# **1.)** OVERVIEW/CONTEXT

---
Text goes here....

---

<figure style="text-align:center;">
    <img src="notebook_images/umn_banner.png" alt="University of Minnesota">
</figure>

---
# <center>**<u>CERT x465-003</u>**: Data Interpretation and Application</center><br><center>Coursework and Assignments</center><br><center>Project Title: ***Deep Poking into Pokémon Data***</center>
<figure style="text-align:center;">
    <img src="notebook_images/pokes_pic.jpg" alt="Pokemon" width="1800">
</figure>

---

# **2.)** *<u>Software/Libraries Herein Implemented</u>*

---
More text goes here....

---

# **3.)** *<u>Assignments</u>*

---

## **3.1)** <u>Assignment 01</u>
- Completed by: Logan R. Sloan<br>

---
### Assignment 01 Details
- Write a short essay that address the following questions:
    - What problem or question are you trying to address?
    - Why are you interested in addressing this?
    - Why is this an important question or problem to address (in general)?
    - Who might your audience(s) be for this investigation?
    - How will the results of this investigation help your audience make decisions or draw conclusions?<br>

---
### Assignment 01 Essay


## **3.2)** <u>Assignment 02-A</u>
- Completed by: Logan R. Sloan<br>

---
### Assignment 02-A Details
- Identify at least one data source that can help answer the question you developed in Module 1.
1. First, do some research on your own to try and find a data source(s).
    - You might start with a simple Google search, you may ask someone  who has knowledge in the area, or you might already know of some candidates.
    - For example, if your question pertains to K-12 education in Minnesota, you might check out the Department of Education's data sources.
 2. After you have finished your search and identified good data source(s), please write a short essay that addresses to the following questions:
    - Describe the data source(s) that you found.
    - How is the data relevant to your question?
    - Where does the data come from?
    - How reliable is the source? Are there any signs of bias?<br>

---
### Assignment 02-A Essay

# **<u>NOTE</u>**: THE DATA SCRAPING SECTION SHOULD BE MOVED TO ASSIGNMENT 02A, SINCE THE POINT OF THE ASSIGNMENT IS TO FIND A DATASET TO USE AND THAT IS WHAT THE SCRAPING, CLEANING, ET CETERA IS DOING....

## **3.3)** <u>Assignment 03</u>
- Completed by: **Logan R. Sloan**<br>

---
### Assignment 03 Details
- Using the data source(s) that you identified, produce a 5-minute video describing what you are seeing in the data and what conclusions the data is pointing towards.
- Make sure you tie it back to the original question you set out to address.
- In the video, please also include a brief discussion of how groups with different perspectives might interpret the data differently.
- Post this video to share with your peers, and provide feedback on the videos of at least two peers.<br>

---

### **3.3.i)** Data Exploration for Assignment 03

---
#### Overview of Methodology to be Used to Completed Assignment 03
- Since I have chosen to create my own dataset, in order to complete *Assignment 03* the dataset must first be compiled together in a clean, usable format.
    - To do this I will, first, scrape multiple websites (and multiple pages on those site) for the base of my dataset.
    - Then I will cleanup this data where necessary, as scraping always involves data that isn't particularly useful in its "raw" state.
    - Lastly, I will expand upon and abstract from the cleaned data to create more columns/categories within the dataset to better aid my analysis of the dataset.
- To explore the dataset, I will use the paradigm of an Exploratory Data Analysis (EDA) and some other Data Science techniques to dig down into the data and see what interesting points in the pokemon dataset we can poke out and prop up for others' viewing.
    - This will include some comparison based observations, basic and more advanced statics calculations, visualizations, as well as some machine learning (clustering) analysis.

---

### **3.3.ii)** Scraping, Cleaning, and Expanding upon the Data to Build Dataset for Analysis

---

### **3.3.ii.0)** Python GPU Acceleration Activation

---

In [1]:
%load_ext cudf.pandas
%load_ext cuml.accel

 ### **3.3.ii.1)** Python Library Imports for Scraping, Cleaning, and Expanding

---

In [2]:
import os
import io
import re
import time
import urllib.request
import urllib.error
import pandas as pd
from bs4 import BeautifulSoup

BASE_URL = 'https://bulbapedia.bulbagarden.net'
DATA_DIR = os.path.expanduser('~/000_Duckspace/WanderduckDevelopment/Ducks/UMN/CERT-x465-003/data/')
_HEADERS = {'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64; rv:132.0) Gecko/20100101 Firefox/132.0'}

 ### **3.3.ii.2)** Base URL to Scrape

In [3]:
 pokemon_ALL_url = "https://bulbapedia.bulbagarden.net/wiki/List_of_Pok%C3%A9mon_by_base_stats_in_Generation_IX"

### **3.3.ii.3)** Initial Scraper (Part 1)
- For part 1 of the initial scraper, most of the information from the tables will be pulled from the base url page (see image below for example).
<br>

<figure style="text-align:center;">
    <img src="notebook_images/initial_scrape_img00.png" alt="Bulbapedia Initial Scrape">
    <figcaption>
        The scraper looks through the HTML (right of image) that laysout the webpage.<br>
        The blue highlight shows that this particular part of the table contains both the Pokémon name and the link to it's individual page (to be used by the next scraper).
    </figcaption>
</figure>

In [4]:
def scrape_pokemon_base_stats(url):
    """
    Scrapes the Bulbapedia base stats table using direct HTTP (no headless browser needed
    since the table is server-rendered HTML).
    Table columns: #, Pokémon (sprite), Pokémon.1 (name), HP, Attack, Defense,
                   Sp. Atk, Sp. Def, Speed, Total, Average
    Returns a list of dicts matching the first 10 columns of pokemon_dataset_MASTER.csv.
    """
    req = urllib.request.Request(url, headers={
        'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64; rv:132.0) Gecko/20100101 Firefox/132.0'
    })
    with urllib.request.urlopen(req, timeout=30) as resp:
        html = resp.read().decode('utf-8')

    dfs = pd.read_html(io.StringIO(html))
    df = next(
        t for t in dfs
        if 'HP' in t.columns and 'Attack' in t.columns
    )

    records = []
    for _, row in df.iterrows():
        try:
            pokedex_num = int(row['#'])
            name        = str(row['Pokémon.1']).strip()   # .1 suffix = name col; plain 'Pokémon' is the sprite (NaN)
            hp          = int(row['HP'])
            attack      = int(row['Attack'])
            defense     = int(row['Defense'])
            sp_atk      = int(row['Sp. Atk'])
            sp_def      = int(row['Sp. Def'])
            speed       = int(row['Speed'])
            total       = int(row['Total'])
            link = f"{BASE_URL}/wiki/{name.replace(' ', '_')}"
        except (ValueError, TypeError):
            continue

        records.append({
            'Pokedex Number': pokedex_num,
            'Pokemon':        name,
            'Link':           link,
            'HP':             hp,
            'Attack':         attack,
            'Defense':        defense,
            'Speed':          speed,
            'Special Attack': sp_atk,
            'Special Defense': sp_def,
            'Stat Total':     total,
        })

    return records


records = scrape_pokemon_base_stats(pokemon_ALL_url)
pokes_scraped = pd.DataFrame(records)
print(f"Scraped {len(pokes_scraped)} rows.")
pokes_scraped.head(10)

Scraped 1239 rows.


,Pokedex Number,Pokemon,Link,HP,Attack,Defense,Speed,Special Attack,Special Defense,Stat Total
0,1,Bulbasaur,https://bulbapedia.bulbagarden.net/wiki/Bulbasaur,45,49,49,45,65,65,318
1,2,Ivysaur,https://bulbapedia.bulbagarden.net/wiki/Ivysaur,60,62,63,60,80,80,405
2,3,Venusaur,https://bulbapedia.bulbagarden.net/wiki/Venusaur,80,82,83,80,100,100,525
3,3,Venusaur Mega Venusaur,https://bulbapedia.bulbagarden.net/wiki/Venusa...,80,100,123,80,122,120,625
4,4,Charmander,https://bulbapedia.bulbagarden.net/wiki/Charma...,39,52,43,65,60,50,309
5,5,Charmeleon,https://bulbapedia.bulbagarden.net/wiki/Charme...,58,64,58,80,80,65,405
6,6,Charizard,https://bulbapedia.bulbagarden.net/wiki/Charizard,78,84,78,100,109,85,534
7,6,Charizard Mega Charizard X,https://bulbapedia.bulbagarden.net/wiki/Chariz...,78,130,111,100,130,85,634
8,6,Charizard Mega Charizard Y,https://bulbapedia.bulbagarden.net/wiki/Chariz...,78,104,78,100,159,115,634
9,7,Squirtle,https://bulbapedia.bulbagarden.net/wiki/Squirtle,44,48,65,43,50,64,314


### **3.3.ii.3.x)** Save Initial Dataset

In [5]:
pokes_scraped.to_csv(DATA_DIR + 'pokemon_base_stats_scraped.csv', index=False)
print(f"Saved {len(pokes_scraped)} rows to {DATA_DIR}pokemon_base_stats_scraped.csv")

Saved 1239 rows to /home/wanderduck/000_Duckspace/WanderduckDevelopment/Ducks/UMN/CERT-x465-003/data/pokemon_base_stats_scraped.csv


### **3.3.ii.3.xx)** Cleanup Pokémon Names Where "Variations" Exist
- If you look at the preview of the initial scraped dataset, you'll see that some of the Pokémon have variations such as "Mega".
- It's a little off topic and overly technical to show how I identified all these (although this section here could probably be defined the same way).
- But the first code block below searches through the names and where variations exist it assigns the variation name to a new column in the dataset named "Variation".
- The second code block then removes the variation excess from the Pokémon's name, so that the standard Pokémon and the variants have the same name (think how you would categorize black dog and yellow dog).

In [6]:
_VAR_PATTERN = re.compile(
    r'Mega.*[A-Za-z]|Alolan.*[A-Za-z]|Partner.*[A-Za-z]|Galarian.*[A-Za-z]'
    r'|Hisuian.*[A-Za-z]|Paldean.*[A-Za-z]+\D|in L.*[A-Za-z]|Primal.*[A-Za-z]'
    r'|[A-Za-z]+.Forme|[A-Za-z]+.Cloak|[A-Za-z]+.Rotom|[A-Za-z]+.Mode'
    r'|[A-Za-z]+.Kyurem|Ash.*[A-Za-z]|Eternal.*[A-Za-z]|[A-Za-z]+.Size'
    r'|[A-Za-z0-9]+%.Forme|Hoopa(?![\s\S]*?Hoopa).*[A-Za-z]'
    r'|D.*[A-Za-z]+.*Necrozma|[A-Za-z]+.Necrozma|Hero.*[A-Za-z]'
    r'|Crowned.*[A-Za-z]|Eternamax.*[A-Za-z]|\b\w+\s+Rider.*[A-Za-z]'
    r'|Bloodmoon.*[A-Za-z]|Male|Female'
)

variations = []
for name in pokes_scraped['Pokemon']:
    match = _VAR_PATTERN.search(name)
    variations.append(match.group() if match else '')

pokes_scraped.insert(10, 'Variation', variations)

n_var = sum(1 for v in variations if v != '')
print(f"Variation column added. {n_var} variants found out of {len(pokes_scraped)} rows.")
pokes_scraped[pokes_scraped['Variation'] != ''].head(10)

Variation column added. 213 variants found out of 1239 rows.


,Pokedex Number,Pokemon,Link,HP,Attack,Defense,Speed,Special Attack,Special Defense,Stat Total,Variation
3,3,Venusaur Mega Venusaur,https://bulbapedia.bulbagarden.net/wiki/Venusa...,80,100,123,80,122,120,625,Mega Venusaur
7,6,Charizard Mega Charizard X,https://bulbapedia.bulbagarden.net/wiki/Chariz...,78,130,111,100,130,85,634,Mega Charizard X
8,6,Charizard Mega Charizard Y,https://bulbapedia.bulbagarden.net/wiki/Chariz...,78,104,78,100,159,115,634,Mega Charizard Y
12,9,Blastoise Mega Blastoise,https://bulbapedia.bulbagarden.net/wiki/Blasto...,79,103,120,78,135,115,630,Mega Blastoise
19,15,Beedrill Mega Beedrill,https://bulbapedia.bulbagarden.net/wiki/Beedri...,65,150,40,145,15,80,495,Mega Beedrill
23,18,Pidgeot Mega Pidgeot,https://bulbapedia.bulbagarden.net/wiki/Pidgeo...,83,80,80,121,135,80,579,Mega Pidgeot
25,19,Rattata Alolan Form,https://bulbapedia.bulbagarden.net/wiki/Rattat...,30,56,35,72,25,35,253,Alolan Form
27,20,Raticate Alolan Form,https://bulbapedia.bulbagarden.net/wiki/Ratica...,75,71,70,77,40,80,413,Alolan Form
33,25,Pikachu Partner Pikachu,https://bulbapedia.bulbagarden.net/wiki/Pikach...,45,80,50,120,75,60,430,Partner Pikachu
35,26,Raichu Alolan Form,https://bulbapedia.bulbagarden.net/wiki/Raichu...,60,85,50,110,95,85,485,Alolan Form


In [7]:
pattern = r'Mega.*[A-Za-z]|Alolan.*[A-Za-z]|Partner.*[A-Za-z]|Galarian.*[A-Za-z]|Hisuian.*[A-Za-z]|Paldean.*[A-Za-z]+\D|in L.*[A-Za-z]|Primal.*[A-Za-z]|[A-Za-z]+.Forme|[A-Za-z]+.Cloak|[A-Za-z]+.Rotom|[A-Za-z]+.Mode|[A-Za-z]+.Kyurem|Ash.*[A-Za-z]|Eternal.*[A-Za-z]|[A-Za-z]+.Size|[A-Za-z0-9]+%.Forme|Hoopa(?![\s\S]*?Hoopa).*[A-Za-z]|D.*[A-Za-z]+.*Necrozma|[A-Za-z]+.Necrozma|Hero.*[A-Za-z]|Crowned.*[A-Za-z]|Eternamax.*[A-Za-z]|\b\w+\s+Rider.*[A-Za-z]|Bloodmoon.*[A-Za-z]|Male|Female'

# Removes the matching parts and clean up whitespace
pokes_scraped['Pokemon'] = pokes_scraped['Pokemon'].str.replace(pattern, '', regex=True).str.strip()

<figure style="text-align:center;">
    <img src="notebook_images/great-success.gif" alt="Great Success! Very Nice!">
    <figcaption>
        Very Nice! Part 1 of the initial scraping is completed!<br>
        Here's a trophy: <span style="display: inline-block; transform: rotate(180deg);">&#x1F3C6;</span>
    </figcaption>
</figure>

---

### **3.3.ii.4)** Initial Scraper (Part 2)
- For part 2 of the initial scraper, the links gathered for each individual Pokémon will all have to be gone through to pull more basic information to expand the dataset.
- Where part one was scraping one webpage, in part 2, one thousand two-hundred thirty-nine (1239 &#x1F480;) webpages will need to be scraped.
- The information to be gathered is Category, Type 1, Type 2, Height (m), and Weight (kg).
- Below is an example from one of these webpages.<br><br>

<figure style="text-align:center;">
    <img src="notebook_images/initial_scrape_img01.png" alt="Second of the Scrapes, Initially....">
    <figcaption>
        <span style="display: inline-block; transform: scale(-1, 1);">&#x1F446;</span>...."Seed Pokémon" under the name is the Category for reference....&#x1F446;
    </figcaption>
</figure>

### **NUMBERS)** Create Column Denoting Pokémon Generation
- There are nine generations of Pokémon apparently, and it might be worthwhile to examine changes over the generations
- The "Pokedex Number" column is the best way to do this because variants have the same number as their standard siblings (as opposed to index values or something else).
- The first line of code creates an empty column named "Generation", the all the following lines fill-in which generation number.

In [8]:
pokes_scraped['Generation'] = np.nan
pokes_scraped.loc[pokes_scraped['Pokedex Number'] <= 151, 'Generation'] = 1
pokes_scraped.loc[pokes_scraped['Pokedex Number'] > 151, 'Generation'] = 2
pokes_scraped.loc[pokes_scraped['Pokedex Number'] > 251, 'Generation'] = 3
pokes_scraped.loc[pokes_scraped['Pokedex Number'] > 386, 'Generation'] = 4
pokes_scraped.loc[pokes_scraped['Pokedex Number'] > 493, 'Generation'] = 5
pokes_scraped.loc[pokes_scraped['Pokedex Number'] > 649, 'Generation'] = 6
pokes_scraped.loc[pokes_scraped['Pokedex Number'] > 721, 'Generation'] = 7
pokes_scraped.loc[pokes_scraped['Pokedex Number'] > 809, 'Generation'] = 8
pokes_scraped.loc[pokes_scraped['Pokedex Number'] > 905, 'Generation'] = 9

NameError: name 'np' is not defined

---

### **NUMBERS)** Pokémon Evolution Stage Scraper
- Pokémon evolve as they get stronger or something: People train them and they grow into more ridiculous versions of themselves and this is termed "evolution" in the lore, I think.
- Tracking how the Pokémon change as they evolve will add great insight into the analysis.
- Pokémon, as far as I can tell, either have a two-stage evolution (first stage changes into second/final) or have a three-stage evolution (where the first stage changes into second/intermediate and second into third/final).
- Also, it should be noted that not all Pokémon evolve; as you will see in the analysis, only about 1/3 of Pokémon evolve into other Pokémon.

### **NUMBERS)** Evolution URLs
- I found separate pages on the same website that organize the two-stage evolutions from the three-stage
- This makes the scraping much easier, but it does require rolling through two webpages instead of a single page, but this written to be extremely fast, so no problem....

In [ ]:
evo3_url = 'https://pokemon.fandom.com/wiki/List_of_Pok%C3%A9mon_by_three-stage_evolution'
evo2_url = 'https://pokemon.fandom.com/wiki/List_of_Pok%C3%A9mon_by_two-stage_evolution'

### **NUMBERS)** Evolution Stage Scraper

In [ ]:
DATA_DIR = os.path.expanduser('~/000_Duckspace/WanderduckDevelopment/Ducks/UMN/CERT-x465-003/data/')


async def get_pokes_rows_pw(page, url):
    """Navigate to url and return all non-header rows from .prettytable tables."""
    await page.goto(url, wait_until='domcontentloaded')
    await page.wait_for_selector('.mw-parser-output .prettytable')
    rows = []
    for table in await page.locator('.mw-parser-output .prettytable').all():
        for i, row in enumerate(await table.locator('tr').all()):
            if i > 0:
                rows.append(row)
    return rows


async def get_name(cell):
    """Return the Pokémon name from a table cell (second anchor if present, else first, else cell text)."""
    anchors = await cell.locator('a').all()
    if len(anchors) >= 2:
        return await anchors[1].inner_text()
    elif len(anchors) == 1:
        return await anchors[0].inner_text()
    return await cell.inner_text()


async def pokes_parser1_pw(row):
    """Parse a three-stage evolution table row → dict with ev1/ev2/ev3 keys."""
    cells = await row.locator('td').all()
    n = len(cells)
    if n == 3:
        return {'ev1': await get_name(cells[0]), 'ev2': await get_name(cells[1]), 'ev3': await get_name(cells[2])}
    elif n == 2:
        return {'ev2': await get_name(cells[0]), 'ev3': await get_name(cells[1])}
    else:
        return {'ev3': await get_name(cells[0])}


async def pokes_parser2_pw(row):
    """Parse a two-stage evolution table row → dict with ev1/ev2 keys."""
    cells = await row.locator('td').all()
    n = len(cells)
    if n == 2:
        first_anchors = await cells[0].locator('a').all()
        ev1 = await first_anchors[1].inner_text() if len(first_anchors) >= 2 else await cells[0].inner_text()
        ev2 = await get_name(cells[1])
        return {'ev1': ev1, 'ev2': ev2}
    else:
        cell_anchors = await cells[0].locator('a').all()
        ev2 = await cell_anchors[1].inner_text() if len(cell_anchors) >= 2 else await cells[0].inner_text()
        return {'ev2': ev2}


async def main():
    async with async_playwright() as pw:
        browser = await pw.firefox.launch(headless=True)
        page = await browser.new_page()

        rows3 = await get_pokes_rows_pw(page, evo3_url)
        evo3_data = [await pokes_parser1_pw(r) for r in rows3]
        pd.DataFrame(evo3_data).to_csv(DATA_DIR + 'evo3.csv', index=False)

        rows2 = await get_pokes_rows_pw(page, evo2_url)
        evo2_data = [await pokes_parser2_pw(r) for r in rows2]
        pd.DataFrame(evo2_data).to_csv(DATA_DIR + 'evo2.csv', index=False)

        await browser.close()

await main()
print("Scraping complete. CSVs saved to data directory.")

### **NUMBERS)** Cleaning-up the Evolution Data {***<u>WARNING</u>***: Long-winded explanation of what's happening below--PLEASE SKIP!!!}
- The scraper above creates two seperate `.csv` files (one for each url).
- So, the two will need to be merged together and then merged into the larger, main dataset, but before that can be done, the evolution stage data needs to be turned into integers.
- The reason why the data wasn't scraped as integers is because Python doesn't allow naming variables with only numbers.
- Python variables must start with either an uppercase or lowercase letter [a-z, A-Z] or an underscore [ _ ].
- And when pulling information from a website like this, the data must be stored as a variable to be compiled together.
- First, the data from each website is compiled into what's called a dictionary.
    - Dictionaries work sort of like cabinets:
        1. There are *unique* "keys" that function like each cabinet box in your kitchen [you cannot have keys of the same value just like every cabinet is a totally separate space], and
        2. There are *non-unique* "values" within those keys that can repeat (like having two of the same pots and whatnot).
- Second, the two separate dictionaries are saved as `.csv` files.
    - A `.csv` is used for three reasons:
        1.  In case something messes up in the integer conversion or merging, we won't need to run the scraper again (and can simply work on fixing our crap code).
        2. because `.csv` files can easily loaded into what are called DataFrames, and these are very easy to manipulate and merge together.
        3. There are certain Pokémon that have the ability to evolve into more than one, singular Pokémon in the next stage, and the way the website's table is setup, I could not find a good work around to match these separate evolutions.
            - [Reference the parenthetical from #1]--This being the case and my code being crap, a `.csv` file is human-readable and easy to manually edit.
            - If you *must* know, it was easy to edit because the Pokémon missing their evolution stages are directly below the stages they are missing, so I just ran a script that found blanks, moved a row up, copied the correct number of values, and pasted them in the blanks.
            - See the picture below to better understand how the table is set up: because of how HTML functions, the Pokémon "Bellossom" is technically on a row by itself, where as "Vileplume" is rowed with the first two.
<figure style="text-align:center;">
    <img src="notebook_images/evo_scrape_img00.png" alt="Gloom of HTML">
    <figcaption>"Gloom" is my natural habitat....
        <span style="display: inline-block; transform: scale(1.5, -1);">&#x1F984;</span>
    </figcaption>
</figure>
<br>

---
- This all means that cleaning the code will work in this order:
    1. Manually filling in the missing Pokémon for the edge cases.
    2. Loading in the separate `.csv` files.
    3. Merging, one-at-a-time, the two `.csv` files into the main dataset
    4. Creating two new columns in the main dataset from the non-integer data (which works to convert it to integers in the process).
        1. <u>"Evolution Stage"</u>: This is the integer value of the Pokémon's place in the evolution chain (1, 2, or 3--and 0 for Pokémon that don't evolve).
        2. <u>"Evolves"</u>: This is a boolean value (i.e. True or False) that tells us whether a Pokémon evolves or not. This is to make it easier to analyze non-evolving Pokémon vs. evolving Pokémon.

In [ ]:
evo3 = pd.read_csv(DATA_DIR + 'evo3.csv')
evo2 = pd.read_csv(DATA_DIR + 'evo2.csv')


# Merge three-stage evolution data into df
df['ev1'] = df['Pokemon'].isin(evo3['ev1'])
df['ev2'] = df['Pokemon'].isin(evo3['ev2'])
df['ev3'] = df['Pokemon'].isin(evo3['ev3'])

df.loc[df['ev1'] == True, ['Evolution Stage', 'Evolves']] = [1, True]
df.loc[df['ev2'] == True, ['Evolution Stage', 'Evolves']] = [2, True]
df.loc[df['ev3'] == True, ['Evolution Stage', 'Evolves']] = [3, False]

df.drop(['ev1', 'ev2', 'ev3'], axis=1, inplace=True)

# Merge two-stage evolution data into df
df['ev1'] = df['Pokemon'].isin(evo2['ev1'])
df['ev2'] = df['Pokemon'].isin(evo2['ev2'])

df.loc[df['ev1'] == True, ['Evolution Stage', 'Evolves']] = [1, True]
df.loc[df['ev2'] == True, ['Evolution Stage', 'Evolves']] = [2, False]

df.drop(['ev1', 'ev2'], axis=1, inplace=True)

# Fill remaining NaN evolution stages (non-evolving Pokémon)
df.loc[df['Evolution Stage'].isna(), ['Evolution Stage', 'Evolves']] = [0, False]

print("Cleaning and merging complete.")

### **NUMBERS)** Probably a Good Time to Save Our Progress

In [ ]:
pokes_scraped.to_csv(('~/000_Duckspace/WanderduckDevelopment/Ducks/UMN/CERT-x465-003/data/pokemon_dataset_MASTER.csv', index=False))

###

### **3.3.iii)** Exploratory Data Analysis (EDA)

---

#### EDA Overview
- This section contains the full Exploratory Data Analysis for Assignment 03.
- The dataset used is `pokemon_dataset_MASTER.csv` (1,231 rows × 55 columns), custom-built from scraped Bulbapedia and Pokémon Fandom Wiki data.
- Eight thematic questions are explored, followed by an unsupervised machine learning clustering capstone.

In [1]:
import os
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from scipy.stats import gaussian_kde as _gaussian_kde
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.metrics import silhouette_score
import warnings
warnings.filterwarnings('ignore')

STAT_COLS   = ['HP', 'Attack', 'Defense', 'Speed', 'Special Attack', 'Special Defense']
ROLE_ORDER  = ['Physical Sweeper', 'Special Sweeper', 'Mixed Attacker',
               'Balanced', 'Tank', 'Physical Wall', 'Special Wall']
PCT_COLS    = ['HP_pct', 'Attack_pct', 'Defense_pct',
               'Speed_pct', 'Special_Attack_pct', 'Special_Defense_pct']
ZSCORE_COLS = ['HP_zscore', 'Attack_zscore', 'Defense_zscore',
               'Speed_zscore', 'Special_Attack_zscore', 'Special_Defense_zscore']

DATA_DIR = os.path.expanduser(
    '~/000_Duckspace/WanderduckDevelopment/Ducks/UMN/CERT-x465-003/data/'
)
pokes = pd.read_csv(DATA_DIR + 'pokemon_dataset_MASTER.csv')

# Chart defaults applied to every figure
_H = 1200
_W = 2000
_T = 'plotly_dark'

In [2]:
# --- Official Pokémon Type Color Palette ---
TYPE_COLORS = {
    'Normal': '#A8A77A',   'Fire': '#EE8130',    'Water': '#6390F0',
    'Electric': '#F7D02C', 'Grass': '#7AC74C',   'Ice': '#96D9D6',
    'Fighting': '#C22E28', 'Poison': '#A33EA1',  'Ground': '#E2BF65',
    'Flying': '#A98FF3',   'Psychic': '#F95587', 'Bug': '#A6B91A',
    'Rock': '#B6A136',     'Ghost': '#735797',   'Dragon': '#6F35FC',
    'Dark': '#705746',     'Steel': '#B7B7CE',   'Fairy': '#D685AD',
}

def _hex_scale(hex_color, factor):
    """factor < 1 = darken, factor > 1 = lighten toward white."""
    h = hex_color.lstrip('#')
    r, g, b = int(h[0:2], 16), int(h[2:4], 16), int(h[4:6], 16)
    if factor > 1:
        f = factor - 1
        r = int(r + (255 - r) * f)
        g = int(g + (255 - g) * f)
        b = int(b + (255 - b) * f)
    else:
        r, g, b = int(r * factor), int(g * factor), int(b * factor)
    return '#{:02x}{:02x}{:02x}'.format(min(r, 255), min(g, 255), min(b, 255))

TYPE_COLORS_LIGHT = {t: _hex_scale(c, 1.45) for t, c in TYPE_COLORS.items()}
TYPE_COLORS_DARK  = {t: _hex_scale(c, 0.60) for t, c in TYPE_COLORS.items()}

ROLE_COLORS   = ['#66C2A5', '#FC8D62', '#8DA0CB', '#E78AC3', '#A6D854', '#FFD92F', '#E5C494']
ROLE_COLOR_MAP = dict(zip(ROLE_ORDER, ROLE_COLORS))

def stats_note(fig, text, y_pos=-0.10, bottom_margin=220):
    """Append a centered stats block below the chart area."""
    fig.update_layout(margin=dict(b=bottom_margin))
    fig.add_annotation(
        text=text, xref='paper', yref='paper',
        x=0.5, y=y_pos, showarrow=False,
        font=dict(size=16, color='#cccccc'), align='center',
        bgcolor='rgba(30,30,30,0.75)',
        bordercolor='rgba(200,200,200,0.35)', borderwidth=1, borderpad=10,
    )
    return fig

print(f"TYPE_COLORS loaded: {len(TYPE_COLORS)} types")

TYPE_COLORS loaded: 18 types


In [3]:
# --- Dataset Verification ---
assert pokes.shape == (1231, 55), f"Expected (1231, 55), got {pokes.shape}"
assert 'BMI' in pokes.columns,             "Missing column: BMI"
assert 'Role' in pokes.columns,            "Missing column: Role"
assert 'Stat Std Dev' in pokes.columns,    "Missing column: Stat Std Dev"
assert 'Legendary' in pokes.columns,       "Missing column: Legendary"
assert 'Evolution Stage' in pokes.columns, "Missing column: Evolution Stage"
for col in ZSCORE_COLS:
    assert col in pokes.columns, f"Missing z-score column: {col}"

print(f"Dataset loaded: {pokes.shape[0]} rows × {pokes.shape[1]} columns")
print(f"\nRole distribution:\n{pokes['Role'].value_counts().to_string()}")
print(f"\nMissing values in key columns:\n"
      f"{pokes[['BMI','Role','Stat Std Dev','Legendary','Evolution Stage']].isna().sum().to_string()}")

Dataset loaded: 1231 rows × 55 columns

Role distribution:
Role
Tank                389
Balanced            355
Special Sweeper     124
Mixed Attacker      100
Physical Sweeper     97
Physical Wall        85
Special Wall         81

Missing values in key columns:
BMI                0
Role               0
Stat Std Dev       0
Legendary          0
Evolution Stage    0


### **3.3.iii.1)** Section 1: Body Composition and Combat Role

---
**Question:** Does a Pokémon's physical body shape (BMI, height, weight) predict how it fights?

In [4]:
# --- 1a: BMI Distribution ---
bmi_all  = pokes['BMI'].clip(upper=200)
bmi_filt = pokes.loc[pokes['BMI'] < 200, 'BMI']

fig = make_subplots(rows=1, cols=2,
                    subplot_titles=['BMI Distribution (clipped at 200)',
                                    'BMI Distribution (BMI < 200)'])

for col_idx, (vals, clr) in enumerate(
        [(bmi_all, '#5B9BD5'), (bmi_filt, '#F28B82')], 1):
    fig.add_trace(go.Histogram(
        x=vals, nbinsx=60,
        marker_color=clr, opacity=0.9,
        marker_line_width=0.77, marker_line_color='rgba(0,0,0,0.4)',
        showlegend=False,
    ), row=1, col=col_idx)
    kde = _gaussian_kde(vals.dropna())
    x_kde = np.linspace(float(vals.min()), float(vals.max()), 300)
    y_kde = kde(x_kde)
    bin_w = (float(vals.max()) - float(vals.min())) / 60
    fig.add_trace(go.Scatter(
        x=x_kde, y=y_kde * len(vals) * bin_w,
        mode='lines', line=dict(color='white', width=2),
        showlegend=False,
    ), row=1, col=col_idx)

fig.update_layout(
    title=dict(text='<b>Pokémon BMI Distribution</b>', x=0.5, font=dict(size=20)),
    template=_T, height=_H, width=_W, bargap=0,
)

top10 = pokes.nlargest(10, 'BMI')[['Pokemon', 'Type 1', 'Height (m)', 'Weight (kg)', 'BMI', 'Role']]
note = '<b>Top 10 Highest-BMI Pokémon:</b><br>' + '<br>'.join(
    f"{r.Pokemon} ({r['Type 1']})  BMI={r.BMI:.1f}  {r.Role}"
    for _, r in top10.iterrows()
)
stats_note(fig, note)
fig.show()

In [5]:
# --- 1b: BMI by Combat Role ---
bmi_data = pokes[pokes['BMI'] < 200].copy()

fig = go.Figure()
for role in ROLE_ORDER:
    subset = bmi_data[bmi_data['Role'] == role]['BMI']
    fig.add_trace(go.Violin(
        y=subset, name=role,
        fillcolor=ROLE_COLOR_MAP.get(role, '#888888'),
        line=dict(color='white', width=3),
        opacity=0.75, box_visible=True,
        meanline_visible=True, points=False,
    ))

fig.update_layout(
    title=dict(text='<b>BMI Distribution by Combat Role</b>', x=0.5, font=dict(size=20)),
    xaxis_title='Combat Role', yaxis_title='BMI',
    template=_T, height=_H, width=_W,
    violinmode='overlay', showlegend=False,
)

role_stats = bmi_data.groupby('Role')['BMI'].agg(['median', 'mean']).reindex(ROLE_ORDER)
note = '  |  '.join(
    f"{r}: median={row['median']:.1f}" for r, row in role_stats.iterrows()
)
stats_note(fig, f'<b>Median BMI by Role:</b><br>{note}')
fig.show()


In [6]:
# --- 1c: Height vs Weight scatter (log scale, interactive) ---
fig = px.scatter(
    pokes[pokes['BMI'] < 500],
    x='Height (m)', y='Weight (kg)',
    color='Type 1',
    color_discrete_map=TYPE_COLORS,
    hover_name='Pokemon',
    hover_data=['BMI', 'Role', 'Generation'],
    log_x=True, log_y=True,
    title='<b>Height vs. Weight by Primary Type (log-log scale)</b>',
    template=_T, opacity=0.7,
    height=_H, width=_W,
)

fig.update_traces(marker=dict(size=17))

heaviest = pokes.nlargest(5, 'Weight (kg)')[['Pokemon', 'Type 1', 'Weight (kg)', 'Height (m)', 'BMI']]
note = '  |  '.join(
    f"{r.Pokemon}: {r['Weight (kg)']}kg, {r['Height (m)']}m, BMI={r.BMI:.1f}"
    for _, r in heaviest.iterrows()
)
stats_note(fig, f'<b>Top 5 Heaviest:</b><br>{note}')
fig.show()


In [7]:
# --- 1d: Median BMI by Generation ---
gen_bmi = pokes.groupby('Generation')['BMI'].median().reset_index()

fig = px.bar(
    gen_bmi, x='Generation', y='BMI',
    color='BMI', color_continuous_scale='Viridis',
    title='<b>Median Pokémon BMI by Generation</b>',
    labels={'BMI': 'Median BMI'},
    template=_T, height=_H, width=_W,
)
fig.update_coloraxes(showscale=False)
fig.update_layout(xaxis=dict(tickmode='linear', tick0=1, dtick=1))

note = '  |  '.join(f"Gen {int(r.Generation)}: {r.BMI:.1f}" for _, r in gen_bmi.iterrows())
stats_note(fig, f'<b>Median BMI per Generation:</b><br>{note}')
fig.show()


### **3.3.iii.2)** Section 2: Type Shapes Stat Identity, Not Just Stat Total

---
**Question:** Does a Pokémon's primary type determine its fighting personality more than its raw power level?

In [8]:
# --- 2a: Combat ratio profiles by Type 1 ---
type_ps = pokes.groupby('Type 1')['Physical/Special'].median().sort_values()
type_od = pokes.groupby('Type 1')['Offensive/Defensive'].median().sort_values()

fig = make_subplots(rows=1, cols=2,
                    subplot_titles=['Median Physical/Special Ratio by Type',
                                    'Median Offensive/Defensive Ratio by Type'])

fig.add_trace(go.Bar(
    y=type_ps.index.tolist(), x=type_ps.values, orientation='h',
    marker_color=[TYPE_COLORS.get(t, '#888') for t in type_ps.index],
    showlegend=False,
), row=1, col=1)
fig.add_vline(x=1.0, line_dash='dash', line_color='red', row=1, col=1)

fig.add_trace(go.Bar(
    y=type_od.index.tolist(), x=type_od.values, orientation='h',
    marker_color=[TYPE_COLORS.get(t, '#888') for t in type_od.index],
    showlegend=False,
), row=1, col=2)
fig.add_vline(x=1.0, line_dash='dash', line_color='red', row=1, col=2)

fig.update_layout(
    title=dict(text='<b>Combat Personality by Primary Type</b>', x=0.5, font=dict(size=20)),
    template=_T, height=_H, width=_W,
)
fig.update_xaxes(title_text='Ratio  (dashed = balanced at 1.0)')

note = (
    f'Most physical: {type_ps.idxmax()} ({type_ps.max():.2f})  |  '
    f'Most special: {type_ps.idxmin()} ({type_ps.min():.2f})<br>'
    f'Most offensive: {type_od.idxmax()} ({type_od.max():.2f})  |  '
    f'Most defensive: {type_od.idxmin()} ({type_od.min():.2f})'
)
stats_note(fig, note)
fig.show()


In [9]:
# --- 2b: Stat fingerprint radar charts (all types) ---
ALL_TYPES  = list(TYPE_COLORS.keys())
STAT_THETA = ['HP', 'Attack', 'Defense', 'Speed', 'Special Attack', 'Special Defense']

fig = make_subplots(
    rows=6, cols=3,
    specs=[[{'type': 'polar'}] * 3] * 6,
    subplot_titles=[f'<b>{t}</b>' for t in ALL_TYPES],
    vertical_spacing=0.05, horizontal_spacing=0.06,
)

polar_max = int(np.ceil(pokes[STAT_THETA].max().max() / 10) * 10)

for i, ptype in enumerate(ALL_TYPES):
    row, col = divmod(i, 3)
    row += 1; col += 1
    means = pokes[pokes['Type 1'] == ptype][STAT_THETA].mean().tolist()
    clr   = TYPE_COLORS.get(ptype, '#888888')
    fig.add_trace(go.Scatterpolar(
        r=means + [means[0]],
        theta=STAT_THETA + [STAT_THETA[0]],
        fill='toself', fillcolor=clr, opacity=0.5,
        line=dict(color=clr, width=2),
        name=ptype, showlegend=False,
    ), row=row, col=col)
    key = 'polar' if i == 0 else f'polar{i + 1}'
    fig.update_layout(**{key: dict(radialaxis=dict(range=[0, polar_max], tickfont=dict(size=8)))})

fig.update_layout(
    title=dict(text='<b>Stat Profile "Fingerprint" by Type</b>', x=0.5, font=dict(size=20)),
    template=_T, height=_H * 3, width=_W,
)
fig.show()

In [10]:
# --- 2c: Does dual typing affect stat profile shape? ---
pokes['Type Label'] = pokes['Dual Type'].map({True: 'Dual Type', False: 'Mono Type'})

fig = make_subplots(rows=1, cols=2,
                    subplot_titles=['Physical/Special by Type Count',
                                    'Offensive/Defensive by Type Count'])

for col_idx, metric in enumerate(['Physical/Special', 'Offensive/Defensive'], 1):
    for label, clr in [('Mono Type', '#5B9BD5'), ('Dual Type', '#F28B82')]:
        subset = pokes[pokes['Type Label'] == label][metric]
        fig.add_trace(go.Violin(
            y=subset, name=label,
            fillcolor=clr,
            line=dict(color='white', width=3),
            opacity=0.75, box_visible=True,
            meanline_visible=True, points=False,
            legendgroup=label,
            showlegend=(col_idx == 1),
        ), row=1, col=col_idx)
    fig.add_hline(y=1.0, line_dash='dash', line_color='rgba(255,255,0,0.5)', row=1, col=col_idx)

fig.update_layout(
    title=dict(text='<b>Does Dual Typing Affect Stat Identity?</b>', x=0.5, font=dict(size=20)),
    template=_T, height=_H, width=_W,
    violinmode='group',
    violingap=0.1, violingroupgap=0.05,
)

medians = pokes.groupby('Type Label')[['Physical/Special', 'Offensive/Defensive']].median().round(3)
note = (
    f"<b>Median ratios:</b><br>"
    f"Mono Type — Phys/Spec: {medians.loc['Mono Type','Physical/Special']}  |  "
    f"Off/Def: {medians.loc['Mono Type','Offensive/Defensive']}<br>"
    f"Dual Type — Phys/Spec: {medians.loc['Dual Type','Physical/Special']}  |  "
    f"Off/Def: {medians.loc['Dual Type','Offensive/Defensive']}"
)
stats_note(fig, note)
fig.show()

In [11]:
# --- 2d: Stat fingerprint ScatterPolar — Primary vs. Secondary type ---
# Type 2 drawn FIRST (behind): darker shade, opacity=0.72, thick solid white border
# Type 1 drawn SECOND (on top): lighter shade, opacity=0.45, thin dotted border
STAT_THETA = ['HP', 'Attack', 'Defense', 'Speed', 'Special Attack', 'Special Defense']
TYPE_ORDER = list(TYPE_COLORS.keys())

all_means = []
for t in TYPE_ORDER:
    for subset in [pokes[pokes['Type 1'] == t], pokes[pokes['Type 2'] == t]]:
        if len(subset) > 0:
            all_means.extend(subset[STAT_THETA].mean().tolist())
POLAR_MAX = int(np.ceil(max(all_means) / 10) * 10)

ROWS, COLS = 6, 3
fig = make_subplots(
    rows=ROWS, cols=COLS,
    specs=[[{'type': 'polar'}] * COLS] * ROWS,
    subplot_titles=[f'<b>{t}</b>' for t in TYPE_ORDER],
    vertical_spacing=0.04,
    horizontal_spacing=0.04,
)

for idx, t in enumerate(TYPE_ORDER):
    row = idx // COLS + 1
    col = idx % COLS + 1

    # --- Type 2 FIRST (drawn behind): dark shade, high opacity, thick border ---
    t2 = pokes[pokes['Type 2'] == t]
    if len(t2) > 0:
        fig.add_trace(go.Scatterpolar(
            r=t2[STAT_THETA].mean().tolist(),
            theta=STAT_THETA,
            fill='toself',
            fillcolor=TYPE_COLORS_DARK[t],
            opacity=0.72,
            line=dict(color='white', width=3),
            name='Secondary (Type 2)',
            legendgroup='type2',
            showlegend=(idx == 0),
        ), row, col)

    # --- Type 1 SECOND (drawn on top): light shade, lower opacity, thin dotted ---
    t1 = pokes[pokes['Type 1'] == t]
    if len(t1) > 0:
        fig.add_trace(go.Scatterpolar(
            r=t1[STAT_THETA].mean().tolist(),
            theta=STAT_THETA,
            fill='toself',
            fillcolor=TYPE_COLORS_LIGHT[t],
            opacity=0.45,
            line=dict(color='rgba(255,255,255,0.6)', width=1.5, dash='dot'),
            name='Primary (Type 1)',
            legendgroup='type1',
            showlegend=(idx == 0),
        ), row, col)

polar_kwargs = {}
for i in range(1, 19):
    key = 'polar' if i == 1 else f'polar{i}'
    polar_kwargs[key] = dict(radialaxis=dict(
        range=[0, POLAR_MAX], showticklabels=True, tickfont=dict(size=10),
    ))

fig.update_layout(
    **polar_kwargs,
    title=dict(
        text=(
            '<b>Mean Stat Profile by Type: Primary vs. Secondary</b><br><br>'
            '<sup>Dark solid fill = Secondary type (Type 2) drawn first  ·  '
            'Light dotted = Primary type (Type 1) on top</sup>'
        ),
        x=0.5, font=dict(size=18),
    ),
    height=_H * 2,
    width=_W,
    template=_T,
    legend=dict(orientation='h', y=1.01, x=0.5, xanchor='center'),
    font=dict(size=14),
)
fig.show()

### **3.3.iii.3)** Section 3: The Specialization Premium

---
**Question:** Do more powerful Pokémon pay for their power by becoming one-dimensional?

In [12]:
# --- 3a: Stat Std Dev vs. Stat Total ---
tier_clr_map = {'Low': '#91CF60', 'Mid': '#FEE08B', 'High': '#FC8D59', 'Very High': '#D73027'}

fig = px.scatter(
    pokes.dropna(subset=['Stat Total', 'Stat Std Dev', 'Stat Tier']),
    x='Stat Total', y='Stat Std Dev',
    color='Stat Tier',
    color_discrete_map=tier_clr_map,
    hover_name='Pokemon',
    hover_data=['Role', 'Type 1', 'Legendary'],
    trendline='ols',
    trendline_scope='overall',
    trendline_color_override='red',
    title='<b>Does Higher Power = Greater Specialization?</b>',
    template=_T, height=_H, width=_W,
    opacity=0.55,
    category_orders={'Stat Tier': ['Low', 'Mid', 'High', 'Very High']},
)
fig.update_layout(yaxis_title='Stat Std Dev  (higher = more specialized)')

fig.update_traces(marker=dict(size=17))

r = np.corrcoef(pokes['Stat Total'].fillna(0), pokes['Stat Std Dev'].fillna(0))[0, 1]
stats_note(fig, f'<b>Pearson r (Stat Total vs. Stat Std Dev):</b>  {r:.4f}')
fig.show()


In [13]:
# --- 3b: Stat Std Dev by Stat Tier and by Role ---
tier_order   = ['Low', 'Mid', 'High', 'Very High']
tier_clr_map = {'Low': '#91CF60', 'Mid': '#FEE08B', 'High': '#FC8D59', 'Very High': '#D73027'}

fig = make_subplots(rows=1, cols=2,
                    subplot_titles=['Stat Std Dev by Stat Tier', 'Stat Std Dev by Combat Role'])

for tier in tier_order:
    fig.add_trace(go.Box(
        y=pokes[pokes['Stat Tier'] == tier]['Stat Std Dev'],
        name=tier, marker_color=tier_clr_map[tier],
        boxmean='sd', showlegend=False,
    ), row=1, col=1)

for role in ROLE_ORDER:
    fig.add_trace(go.Violin(
        y=pokes[pokes['Role'] == role]['Stat Std Dev'],
        name=role, fillcolor=ROLE_COLOR_MAP.get(role, '#888'),
        line=dict(color='white', width=3),
        opacity=0.75,
        box_visible=True, meanline_visible=True, points=False,
        showlegend=False,
    ), row=1, col=2)

fig.update_layout(
    title=dict(text='<b>The Specialization Premium</b>', x=0.5, font=dict(size=20)),
    yaxis_title='Stat Std Dev', yaxis2_title='Stat Std Dev',
    template=_T, height=_H, width=_W,
)

mean_by_tier = pokes.groupby('Stat Tier')['Stat Std Dev'].mean().reindex(tier_order).round(2)
note = '  |  '.join(f"{t}: {v}" for t, v in mean_by_tier.items())
stats_note(fig, f'<b>Mean Stat Std Dev by Tier:</b><br>{note}')
fig.show()


### **3.3.iii.4)** Section 4: The Speed-Bulk Tradeoff Across Generations

---
**Question:** Is "fast = fragile" a real design rule in the data, and has it intensified over time?

In [14]:
# --- 4a: Speed vs. Bulk interactive scatter ---
pokes['Bulk'] = pokes['HP'] + pokes['Defense'] + pokes['Special Defense']
assert pokes['Bulk'].isna().sum() == 0, "Bulk has unexpected NaNs"

fig = px.scatter(
    pokes,
    x='Speed', y='Bulk',
    color='Generation',
    color_continuous_scale='Viridis',
    hover_name='Pokemon',
    hover_data=['Role', 'Stat Total', 'Type 1'],
    trendline='ols',
    title='<b>Speed vs. Bulk (HP + Defense + Sp. Def) — colored by Generation</b>',
    template=_T, opacity=0.6,
    height=_H, width=_W,
)

fig.update_traces(marker=dict(size=17))

r_speed_bulk = pokes[['Speed', 'Bulk']].corr().iloc[0, 1]
stats_note(fig, f'<b>Overall Pearson r (Speed vs. Bulk):</b>  {r_speed_bulk:.4f}')
fig.show()


In [15]:
# --- 4b: Speed-Bulk Pearson r by Generation ---
gen_corrs = (
    pokes.groupby('Generation')
    .apply(lambda df: df['Speed'].corr(df['Bulk']))
    .reset_index(name='Pearson_r')
)

fig = go.Figure()
fig.add_trace(go.Scatter(
    x=gen_corrs['Generation'], y=gen_corrs['Pearson_r'],
    mode='lines+markers',
    line=dict(color='#5B9BD5', width=4),
    marker=dict(size=14, color='#5B9BD5'),
    name='Pearson r',
))
fig.add_hline(y=0, line_dash='dash', line_color='rgba(255,255,255,0.4)')

fig.update_layout(
    title=dict(
        text='<b>Speed–Bulk Correlation by Generation</b><br>'
             '<sup>More Negative = Stronger Fast/Fragile Tradeoff</sup>',
        x=0.5, font=dict(size=20)
    ),
    xaxis=dict(title='Generation', tickmode='linear', tick0=1, dtick=1),
    yaxis_title='Pearson r',
    template=_T, height=_H, width=_W,
)

note = '  |  '.join(f"Gen {int(r.Generation)}: {r.Pearson_r:.3f}" for _, r in gen_corrs.iterrows())
stats_note(fig, f'<b>Pearson r per Generation:</b><br>{note}')
fig.show()


In [16]:
# --- 4c: Speed/Defense ratio distribution by Generation ---
_gen_palette = {str(g): c for g, c in zip(
    sorted(pokes['Generation'].dropna().unique()),
    ['#440154','#46327e','#365c8d','#277f8e','#1fa187','#4ac16d','#a0da39','#fde725','#f5a623']
)}
fig = px.box(
    pokes.assign(Gen=pokes['Generation'].astype(str)),
    x='Gen', y='Speed/Defense',
    color='Gen', color_discrete_map=_gen_palette,
    title='<b>Speed/Defense Ratio Distribution by Generation</b>',
    labels={'Gen': 'Generation'},
    template=_T, height=_H, width=_W,
    points=False,
)
fig.update_layout(showlegend=False)
fig.show()


In [17]:
# --- 4d: Speed Tier × Stat Tier cross-tab heatmap ---
speed_tier_order = ['Slow', 'Medium', 'Fast']
stat_tier_order  = ['Low', 'Mid', 'High', 'Very High']

cross = pd.crosstab(
    pokes['Speed Tier'], pokes['Stat Tier']
).reindex(index=speed_tier_order, columns=stat_tier_order, fill_value=0)

fig = px.imshow(
    cross,
    labels=dict(x='Stat Tier (Overall Power)', y='Speed Tier', color='Count'),
    color_continuous_scale='Blues',
    text_auto=True,
    title='<b>Speed Tier × Stat Tier Cross-Tabulation</b><br>'
          '<sup>Fast Pokémon — are they also high overall power?</sup>',
    template=_T, height=_H, width=_W,
)
fig.update_layout(xaxis=dict(side='bottom'))

spd_num  = pokes['Speed Tier'].map({'Slow': 0, 'Medium': 1, 'Fast': 2}).fillna(0)
stat_num = pokes['Stat Tier'].map({'Low': 0, 'Mid': 1, 'High': 2, 'Very High': 3}).fillna(0)
r_tiers  = spd_num.corr(stat_num)

note = (
    f'<b>Total Pokémon in cross-tab: {int(cross.values.sum())}</b>  |  '
    f'Pearson r (Speed Tier vs. Stat Tier): {r_tiers:.3f}<br>'
    f'Diagonal trend = fast Pokémon tend toward higher overall stat tiers (or vice versa)'
)
stats_note(fig, note)
fig.show()


### **3.3.iii.5)** Section 5: Is Power Creep Asymmetric?

---
**Question:** Has offensive power inflated faster than defensive bulk across nine generations?

In [18]:
# --- 5a: Offensive vs. Defensive totals by generation ---
gen_stats = pokes.groupby('Generation').agg(
    Offensive_Mean=('Offensive Total', 'mean'),
    Defensive_Mean=('Defensive Total', 'mean'),
    StatTotal_Mean=('Stat Total', 'mean'),
    OffDef_Ratio=('Offensive/Defensive', 'mean'),
    **{f'{s}_Mean': (s, 'mean') for s in STAT_COLS}
).reset_index()

assert len(gen_stats) == 9, f"Expected 9 generations, got {len(gen_stats)}"

fig = make_subplots(rows=1, cols=2,
                    subplot_titles=['Mean Offensive vs. Defensive Total by Generation',
                                    'Mean Offensive/Defensive Ratio by Generation'])

fig.add_trace(go.Scatter(
    x=gen_stats['Generation'], y=gen_stats['Offensive_Mean'],
    mode='lines+markers', name='Offensive Total',
    line=dict(color='#F28B82', width=4), marker=dict(size=12),
), row=1, col=1)
fig.add_trace(go.Scatter(
    x=gen_stats['Generation'], y=gen_stats['Defensive_Mean'],
    mode='lines+markers', name='Defensive Total',
    line=dict(color='#5B9BD5', width=4), marker=dict(size=12),
), row=1, col=1)

fig.add_trace(go.Scatter(
    x=gen_stats['Generation'], y=gen_stats['OffDef_Ratio'],
    mode='lines+markers', name='Off/Def Ratio',
    line=dict(color='#FFD700', width=4), marker=dict(size=12),
), row=1, col=2)
fig.add_hline(y=1.0, line_dash='dash', line_color='rgba(255,255,255,0.4)', row=1, col=2)

fig.update_layout(
    title=dict(text='<b>Is Power Creep Asymmetric?</b>', x=0.5, font=dict(size=20)),
    xaxis=dict(title='Generation', tickmode='linear', tick0=1, dtick=1),
    xaxis2=dict(title='Generation', tickmode='linear', tick0=1, dtick=1),
    yaxis_title='Mean Partial Stat Total',
    yaxis2_title='Ratio  (>1 = more offensive)',
    template=_T, height=_H, width=_W,
)

max_off_gen = gen_stats.loc[gen_stats['Offensive_Mean'].idxmax(), 'Generation']
max_def_gen = gen_stats.loc[gen_stats['Defensive_Mean'].idxmax(), 'Generation']
note = (
    f'<b>Peak Offensive Total:</b> Gen {int(max_off_gen)} ({gen_stats["Offensive_Mean"].max():.1f})  |  '
    f'<b>Peak Defensive Total:</b> Gen {int(max_def_gen)} ({gen_stats["Defensive_Mean"].max():.1f})'
)
stats_note(fig, note)
fig.show()


In [19]:
# --- 5b: Which individual stats inflated most? ---
stat_colors = {
    'HP': '#91CF60', 'Attack': '#FC8D59', 'Defense': '#5B9BD5',
    'Speed': '#FFD700', 'Special Attack': '#BF80FF', 'Special Defense': '#7FDBFF',
}

fig = make_subplots(rows=2, cols=3,
                    subplot_titles=[f'Mean {s} by Generation' for s in STAT_COLS],
                    vertical_spacing=0.15, horizontal_spacing=0.08)

for i, stat in enumerate(STAT_COLS):
    row, col = divmod(i, 3)
    row += 1; col += 1
    fig.add_trace(go.Scatter(
        x=gen_stats['Generation'], y=gen_stats[f'{stat}_Mean'],
        mode='lines+markers',
        line=dict(color=stat_colors.get(stat, '#888888'), width=4),
        marker=dict(size=12),
        name=stat, showlegend=False,
    ), row=row, col=col)

fig.update_layout(
    title=dict(text='<b>Which Stats Inflated Most? (Power Creep by Stat)</b>', x=0.5, font=dict(size=20)),
    template=_T, height=_H, width=_W,
)
for i in range(1, 7):
    ax = 'xaxis' if i == 1 else f'xaxis{i}'
    fig.update_layout(**{ax: dict(tickmode='linear', tick0=1, dtick=1)})

delta = gen_stats.set_index('Generation')[[f'{s}_Mean' for s in STAT_COLS]].diff().dropna()
fastest = delta.sum().idxmax().replace('_Mean', '')
stats_note(fig, f'<b>Fastest-growing stat (Gen 1→9 cumulative increase):</b>  {fastest}  ({delta.sum().max():.1f} pts)')
fig.show()


In [20]:
# --- 5c: Stat Total distribution per generation ---
_gen_palette = {str(g): c for g, c in zip(
    sorted(pokes['Generation'].dropna().unique()),
    ['#440154','#46327e','#365c8d','#277f8e','#1fa187','#4ac16d','#a0da39','#fde725','#f5a623']
)}
fig = px.box(
    pokes.assign(Gen=pokes['Generation'].astype(str)),
    x='Gen', y='Stat Total',
    color='Gen', color_discrete_map=_gen_palette,
    title='<b>Stat Total Distribution by Generation</b>',
    labels={'Gen': 'Generation'},
    template=_T, height=_H, width=_W,
    points=False,
)
fig.update_layout(showlegend=False)

note = '  |  '.join(
    f"Gen {int(r.Generation)}: {round(r.StatTotal_Mean, 1)}"
    for _, r in gen_stats.iterrows()
)
stats_note(fig, f'<b>Mean Stat Total per Generation:</b><br>{note}')
fig.show()


### **3.3.iii.6)** Section 6: Is the Legendary Gap Closing?

---
**Question:** Have ordinary Pokémon powered up enough that Legendary status no longer signals exceptional strength?

In [21]:
# --- 6a: Stat Total — Legendary vs Non-Legendary ---
assert pokes['Legendary'].any(), "No Legendary=True rows found"

fig = go.Figure()
for leg, label, clr in [(False, 'Non-Legendary', '#5B9BD5'), (True, 'Legendary', '#FFD700')]:
    fig.add_trace(go.Violin(
        y=pokes[pokes['Legendary'] == leg]['Stat Total'],
        name=label, fillcolor=clr,
        line=dict(color='white', width=3),
        opacity=0.75, box_visible=True, meanline_visible=True, points=False,
    ))

fig.update_layout(
    title=dict(text='<b>Stat Total: Legendary vs. Non-Legendary</b>', x=0.5, font=dict(size=20)),
    yaxis_title='Stat Total',
    template=_T, height=_H, width=_W,
    violinmode='overlay', showlegend=True,
)

summary  = pokes.groupby('Legendary')['Stat Total'].agg(['mean', 'median', 'min', 'max']).round(1)
leg_row  = summary.loc[True]
nleg_row = summary.loc[False]
gap = leg_row['mean'] - nleg_row['mean']
note = (
    f"<b>Non-Legendary:</b> mean={nleg_row['mean']}, median={nleg_row['median']}, "
    f"range=[{nleg_row['min']}, {nleg_row['max']}]<br>"
    f"<b>Legendary:</b> mean={leg_row['mean']}, median={leg_row['median']}, "
    f"range=[{leg_row['min']}, {leg_row['max']}]<br>"
    f"<b>Mean gap (Legendary − Non-Legendary):</b> {gap:.1f}"
)
stats_note(fig, note)
fig.show()


In [22]:
# --- 6b: Legendary gap by generation ---
gen_leg = (
    pokes.groupby(['Generation', 'Legendary'])['Stat Total']
    .mean().unstack('Legendary')
    .rename(columns={False: 'Non-Legendary', True: 'Legendary'})
)
gen_leg['Gap'] = gen_leg['Legendary'] - gen_leg['Non-Legendary']
gen_leg = gen_leg.reset_index()

fig = make_subplots(rows=1, cols=2,
                    subplot_titles=['Mean Stat Total by Generation × Legendary Status',
                                    'Legendary Stat Gap by Generation\n(Legendary − Non-Legendary)'])

for col_key, label, clr in [('Non-Legendary', 'Non-Legendary', '#5B9BD5'),
                              ('Legendary', 'Legendary', '#FFD700')]:
    fig.add_trace(go.Scatter(
        x=gen_leg['Generation'], y=gen_leg[col_key],
        mode='lines+markers', name=label,
        line=dict(color=clr, width=2), marker=dict(size=8),
    ), row=1, col=1)

fig.add_trace(go.Bar(
    x=gen_leg['Generation'], y=gen_leg['Gap'],
    name='Gap', marker_color='#FF8C00',
), row=1, col=2)
fig.add_hline(y=0, line_dash='dash', line_color='rgba(255,255,255,0.4)', row=1, col=2)

fig.update_layout(
    title=dict(text='<b>Is the Legendary Gap Closing?</b>', x=0.5, font=dict(size=20)),
    xaxis=dict(title='Generation', tickmode='linear', tick0=1, dtick=1),
    xaxis2=dict(title='Generation', tickmode='linear', tick0=1, dtick=1),
    yaxis_title='Mean Stat Total',
    yaxis2_title='Stat Total Gap',
    template=_T, height=_H, width=_W,
)

gaps = gen_leg.set_index('Generation')['Gap']
note = '  |  '.join(f"Gen {int(g)}: {v:.1f}" for g, v in gaps.items())
stats_note(fig, f'<b>Legendary gap per generation:</b><br>{note}')
fig.show()


In [23]:
# --- 6c: Where do the top-50 non-legendary Pokémon come from? ---
top50_nonleg = pokes[~pokes['Legendary']].nlargest(50, 'Stat Total')

fig = px.strip(
    top50_nonleg,
    x='Generation', y='Stat Total',
    color='Type 1',
    color_discrete_map=TYPE_COLORS,
    hover_name='Pokemon',
    hover_data=['Role', 'Is Variation', 'Stat Tier'],
    title='<b>Top 50 Non-Legendary Pokémon by Stat Total</b>',
    template=_T, height=_H, width=_W,
)
fig.update_layout(xaxis=dict(tickmode='linear', tick0=1, dtick=1))
fig.update_traces(marker=dict(size=17))

gen_breakdown = top50_nonleg['Generation'].value_counts().sort_index()
note = '  |  '.join(f"Gen {int(g)}: {c}" for g, c in gen_breakdown.items())
stats_note(fig, f'<b>Top-50 non-legendary — generation breakdown:</b><br>{note}')
fig.show()


### **3.3.iii.7)** Section 7: Does Evolution Preserve or Flip Combat Role?

---
**Question:** Is a Stage 1 Pokémon a miniature version of its final form, or does evolution sometimes change the whole game plan?

In [24]:
# --- 7a: Role distribution at each Evolution Stage ---
evo_pokes = pokes[pokes['Evolution Stage'] > 0].copy()
assert len(evo_pokes) > 0, "No evolving Pokémon found"

role_by_stage = (
    evo_pokes.groupby(['Evolution Stage', 'Role'])
    .size().reset_index(name='Count')
)

fig = px.bar(
    role_by_stage, x='Evolution Stage', y='Count',
    color='Role',
    color_discrete_map=ROLE_COLOR_MAP,
    barmode='stack',
    category_orders={'Role': ROLE_ORDER},
    title='<b>Role Distribution by Evolution Stage</b>',
    labels={'Evolution Stage': 'Evolution Stage', 'Count': 'Number of Pokémon'},
    template=_T, height=_H, width=_W,
)
fig.update_layout(xaxis=dict(tickmode='linear'))
fig.show()


In [25]:
# --- 7b: Role transition heatmaps + Sankey diagrams ---
evo2_chains = pd.read_csv(DATA_DIR + 'evo2.csv')
evo3_chains = pd.read_csv(DATA_DIR + 'evo3.csv')

pairs_1_2 = pd.concat([
    evo2_chains.rename(columns={'ev1': 'name_from', 'ev2': 'name_to'}),
    evo3_chains[['ev1', 'ev2']].rename(columns={'ev1': 'name_from', 'ev2': 'name_to'}),
], ignore_index=True).drop_duplicates()

pairs_2_3 = (evo3_chains[['ev2', 'ev3']]
             .rename(columns={'ev2': 'name_from', 'ev3': 'name_to'})
             .drop_duplicates())

base_pokes  = evo_pokes[~evo_pokes['Is Variation']]
role_lookup = base_pokes.drop_duplicates('Pokemon').set_index('Pokemon')['Role']

def build_transition_matrix(pairs_df, role_lookup, role_order):
    df = pairs_df.copy()
    df['From Role'] = df['name_from'].map(role_lookup)
    df['To Role']   = df['name_to'].map(role_lookup)
    df = df.dropna(subset=['From Role', 'To Role'])
    matrix = (df.groupby(['From Role', 'To Role'])
               .size().unstack('To Role').fillna(0))
    return matrix.reindex(index=role_order, columns=role_order, fill_value=0)

trans_1_2 = build_transition_matrix(pairs_1_2, role_lookup, ROLE_ORDER)
trans_2_3 = build_transition_matrix(pairs_2_3, role_lookup, ROLE_ORDER)

# --- Heatmaps ---
fig_heat = make_subplots(rows=1, cols=2,
                         subplot_titles=['Stage 1 → Stage 2 Transitions',
                                         'Stage 2 → Stage 3 Transitions'])
for col_idx, (mat, cmap) in enumerate([(trans_1_2, 'Blues'), (trans_2_3, 'Greens')], 1):
    fig_heat.add_trace(go.Heatmap(
        z=mat.astype(int).values,
        x=mat.columns.tolist(),
        y=mat.index.tolist(),
        text=mat.astype(int).values,
        texttemplate='%{text}',
        colorscale=cmap,
        showscale=False,
    ), row=1, col=col_idx)

fig_heat.update_layout(
    title=dict(text='<b>Does Evolution Flip Combat Role? — Transition Heatmaps</b>',
               x=0.5, font=dict(size=18)),
    template=_T, height=_H, width=_W,
)
fig_heat.update_xaxes(title_text='To Role')
fig_heat.update_yaxes(title_text='From Role')

total_1_2 = trans_1_2.values.sum()
preserved  = sum(trans_1_2.iloc[i, i] for i in range(len(ROLE_ORDER)))
note = (
    f'<b>Stage 1→2 role preservation rate: {preserved/total_1_2:.1%}</b>  |  '
    f'Total Stage 1→2 pairs matched: {int(total_1_2)}'
)
stats_note(fig_heat, note)
fig_heat.show()

# --- Sankey colour helper: blend two hex colours into an rgba string ---
def _blend_hex(c1, c2, opacity=0.6):
    def h2rgb(h):
        h = h.lstrip('#')
        return int(h[0:2], 16), int(h[2:4], 16), int(h[4:6], 16)
    r1, g1, b1 = h2rgb(c1)
    r2, g2, b2 = h2rgb(c2)
    r = (r1 + r2) // 2
    g = (g1 + g2) // 2
    b = (b1 + b2) // 2
    return f'rgba({r},{g},{b},{opacity})'

# --- Sankey helper ---
def make_sankey(trans_mat, title, role_order, role_color_map):
    n = len(role_order)
    node_labels = [f'S1: {r}' for r in role_order] + [f'S2: {r}' for r in role_order]
    node_colors = [role_color_map.get(r, '#888') for r in role_order] * 2
    sources, targets, values, link_colors = [], [], [], []
    for i, from_role in enumerate(role_order):
        for j, to_role in enumerate(role_order):
            v = int(trans_mat.loc[from_role, to_role])
            if v > 0:
                sources.append(i)
                targets.append(n + j)
                values.append(v)
                c1 = role_color_map.get(from_role, '#888888')
                c2 = role_color_map.get(to_role, '#888888')
                link_colors.append(_blend_hex(c1, c2))
    fig = go.Figure(go.Sankey(
        node=dict(pad=15, thickness=20, label=node_labels, color=node_colors),
        link=dict(source=sources, target=targets, value=values, color=link_colors),
    ))
    fig.update_layout(
        title=dict(text=f'<b>{title}</b>', x=0.5, font=dict(size=18)),
        template=_T, height=_H, width=_W, font_size=12,
    )
    return fig

make_sankey(trans_1_2, 'Role Transitions: Stage 1 → Stage 2 (Sankey)',
            ROLE_ORDER, ROLE_COLOR_MAP).show()
make_sankey(trans_2_3, 'Role Transitions: Stage 2 → Stage 3 (Sankey)',
            ROLE_ORDER, ROLE_COLOR_MAP).show()


In [26]:
# --- 7c: Mean stat growth per evolution step ---
stat_by_stage = evo_pokes.groupby('Evolution Stage')[STAT_COLS].mean()
stat_delta    = stat_by_stage.diff().dropna()
stat_delta.index = ['Stage 1→2', 'Stage 2→3']

stat_colors_list = ['#91CF60', '#FC8D59', '#5B9BD5', '#FFD700', '#BF80FF', '#7FDBFF']

fig = make_subplots(rows=1, cols=len(STAT_COLS),
                    subplot_titles=STAT_COLS, shared_yaxes=True)

for i, stat in enumerate(STAT_COLS):
    fig.add_trace(go.Bar(
        x=stat_delta.index.tolist(),
        y=stat_delta[stat].values,
        name=stat,
        marker_color=stat_colors_list[i],
        showlegend=False,
    ), row=1, col=i + 1)

fig.update_layout(
    title=dict(text='<b>Mean Stat Growth Per Evolution Step</b>', x=0.5, font=dict(size=20)),
    yaxis_title='Mean Increase',
    template=_T, height=_H, width=_W,
)

note_lines = ['<b>Mean stat gain per evolution step:</b>']
for idx_name in stat_delta.index:
    vals = '  |  '.join(f"{s}: {stat_delta.loc[idx_name, s]:.1f}" for s in STAT_COLS)
    note_lines.append(f'{idx_name}: {vals}')
stats_note(fig, '<br>'.join(note_lines))
fig.show()


### **3.3.iii.8)** Section 8: ML Capstone — Unsupervised Clustering

---
**Question:** Without using any labels, can the data's own patterns rediscover the role archetypes we defined — and reveal unexpected groupings?

**Method:** K-Means clustering on the six z-score stat columns, visualized via PCA dimensionality reduction.

In [27]:
# --- 8a: Choosing K — Elbow + Silhouette ---
cluster_df = pokes.dropna(subset=ZSCORE_COLS).copy()
X = cluster_df[ZSCORE_COLS].values

assert X.shape[0] > 100, "Too few rows after dropping NaN — check ZSCORE_COLS"

inertias    = []
silhouettes = []
K_range     = range(2, 13)

for k in K_range:
    km  = KMeans(n_clusters=k, random_state=42, n_init=10)
    lbl = km.fit_predict(X)
    inertias.append(km.inertia_)
    silhouettes.append(silhouette_score(X, lbl))

fig = make_subplots(rows=1, cols=2,
                    subplot_titles=['Elbow Method (Inertia vs. K)',
                                    'Silhouette Score vs. K'])

fig.add_trace(go.Scatter(
    x=list(K_range), y=inertias, mode='lines+markers',
    line=dict(color='#5B9BD5', width=4), marker=dict(size=8),
    name='Inertia', showlegend=False,
), row=1, col=1)

fig.add_trace(go.Scatter(
    x=list(K_range), y=silhouettes, mode='lines+markers',
    line=dict(color='#F28B82', width=4), marker=dict(size=8),
    name='Silhouette', showlegend=False,
), row=1, col=2)

fig.update_layout(
    title=dict(text='<b>Choosing the Optimal K</b>', x=0.5, font=dict(size=20)),
    xaxis=dict(title='Number of Clusters (K)', tickmode='linear', tick0=2, dtick=1),
    xaxis2=dict(title='Number of Clusters (K)', tickmode='linear', tick0=2, dtick=1),
    yaxis_title='Inertia',
    yaxis2_title='Silhouette Score',
    template=_T, height=_H, width=_W,
)

best_k = list(K_range)[silhouettes.index(max(silhouettes))]
note = (
    f'<b>Best K by silhouette:</b> {best_k}  (score: {max(silhouettes):.4f})<br>'
    f'<b>K range tested:</b> 2–12  |  K=6 chosen manually for richer cluster structure'
)
stats_note(fig, note)
fig.show()

In [28]:
# --- 8b: Fit K-Means ---
CHOSEN_K = 6  # K=6 for richer cluster exploration (silhouette peaks at K=2)

km_final = KMeans(n_clusters=CHOSEN_K, random_state=42, n_init=10)
cluster_df['Cluster'] = km_final.fit_predict(X)

pca = PCA(n_components=2, random_state=42)
components = pca.fit_transform(X)
cluster_df['PC1'] = components[:, 0]
cluster_df['PC2'] = components[:, 1]

cluster_sizes  = cluster_df['Cluster'].value_counts().sort_index()
cluster_colors = {str(c): ROLE_COLORS[c % len(ROLE_COLORS)]
                  for c in cluster_df['Cluster'].unique()}

fig = px.scatter(
    cluster_df,
    x='PC1', y='PC2',
    color=cluster_df['Cluster'].astype(str),
    color_discrete_map=cluster_colors,
    hover_name='Pokemon',
    hover_data=['Role', 'Type 1', 'Legendary', 'Stat Total', 'Generation'],
    title=f'<b>K-Means Clusters (K={CHOSEN_K}) — PCA Projection</b>',
    template=_T, opacity=0.7,
    height=_H, width=_W,
)
fig.update_layout(legend_title='Cluster')

fig.update_traces(marker=dict(size=17))

sizes_str = '  |  '.join(f"Cluster {c}: {n}" for c, n in cluster_sizes.items())
pca_var   = sum(pca.explained_variance_ratio_[:2])
note = (
    f'<b>Cluster sizes:</b><br>{sizes_str}<br>'
    f'<b>PCA explained variance:</b>  '
    f'PC1={pca.explained_variance_ratio_[0]:.1%}  '
    f'PC2={pca.explained_variance_ratio_[1]:.1%}  '
    f'Total={pca_var:.1%}'
)
stats_note(fig, note)
fig.show()


In [29]:
# --- 8c: How well do data-driven clusters match rule-based roles? ---
cross = pd.crosstab(cluster_df['Cluster'], cluster_df['Role'])[ROLE_ORDER]

fig = px.imshow(
    cross,
    labels=dict(x='Role (rule-based label)',
                y=f'Cluster (data-driven, K={CHOSEN_K})',
                color='Count'),
    color_continuous_scale='Blues',
    text_auto=True,
    title=f'<b>K-Means Cluster vs. Role Label (K={CHOSEN_K})</b>',
    template=_T, height=_H, width=_W,
)
fig.update_layout(xaxis=dict(side='bottom'))

dominant_roles = cross.idxmax(axis=1)
note = '  |  '.join(f"Cluster {c}: dominant={r}" for c, r in dominant_roles.items())
stats_note(fig, f'<b>Dominant role per cluster:</b><br>{note}')
fig.show()


In [30]:
# --- 8d: What does each cluster "look like" statistically? ---
cluster_melted = (
    cluster_df.groupby('Cluster')[STAT_COLS].mean()
    .reset_index()
    .melt(id_vars='Cluster', var_name='Stat', value_name='Mean Value')
)
cluster_melted['Cluster'] = cluster_melted['Cluster'].astype(str)
cluster_colors = {str(c): ROLE_COLORS[int(c) % len(ROLE_COLORS)]
                  for c in cluster_melted['Cluster'].unique()}

fig = px.bar(
    cluster_melted, x='Stat', y='Mean Value',
    color='Cluster', barmode='group',
    color_discrete_map=cluster_colors,
    title=f'<b>Mean Base Stats per Cluster (K={CHOSEN_K})</b>',
    template=_T, height=_H, width=_W,
)
fig.update_layout(xaxis_title='Stat', yaxis_title='Mean Value')

profiles = cluster_df.groupby('Cluster')[STAT_COLS].mean().round(1)
note_lines = ['<b>Cluster mean stat profiles:</b>']
for c, row in profiles.iterrows():
    vals = '  '.join(f"{s}: {row[s]}" for s in STAT_COLS)
    note_lines.append(f'Cluster {c}: {vals}')
stats_note(fig, '<br>'.join(note_lines))
fig.show()


In [31]:
# --- 8e: Notable Pokémon in each cluster ---
top_per_cluster = (
    cluster_df.groupby('Cluster', group_keys=False)
    .apply(lambda df: df.nlargest(5, 'Stat Total'))
    .reset_index(drop=True)
)
cluster_colors = {str(c): ROLE_COLORS[int(c) % len(ROLE_COLORS)]
                  for c in cluster_df['Cluster'].unique()}

fig = px.scatter(
    cluster_df,
    x='PC1', y='PC2',
    color=cluster_df['Cluster'].astype(str),
    color_discrete_map=cluster_colors,
    hover_name='Pokemon',
    hover_data=['Role', 'Type 1', 'Legendary', 'Stat Total'],
    title=(f'<b>Notable Pokémon in Each Cluster (K={CHOSEN_K})</b><br>'
           f'<sup>Top-5 by Stat Total per cluster annotated</sup>'),
    template=_T, opacity=0.4,
    height=_H, width=_W,
)

for _, row in top_per_cluster.iterrows():
    fig.add_annotation(
        x=row['PC1'], y=row['PC2'],
        text=row['Pokemon'],
        showarrow=True, arrowhead=2,
        font=dict(size=9, color='white'),
        bgcolor='rgba(30,30,30,0.7)',
        arrowcolor='rgba(255,255,255,0.5)',
    )

fig.update_traces(marker=dict(size=17))

fig.update_layout(legend_title='Cluster')
fig.show()
